### PR TITLE
Extend plant hydraulics rooting depth to support fields

### LINKED ISSUE #772: Add spatially varying rooting depth to plant hydraulics
- [x]  obtain or create map of rooting depths
- [x]  add the map to ClimaArtifacts
- [x] replace root distribution with rooting depth
- [x] change all calls to `PlantHydraulicsParameters` to support change
- [x] extend rooting depth to support `Union{FT, ClimaCore.Fields.Field}`
- [x]  make relevant additions to tests
- [x]  change land.jl to use the root depth map
- [x] add to NEWS.md and mark as breaking change

### PR DESCRIPTION
**Pull request by imreddyTeja**
## Purpose 
Towards spatially varying canopy parameters
Move the root_distribution function to src
Make rooting_depth, an argument to root_distribution, a parameter of PlantHydraulicsParameters
Make rooting depth support ClimaCore fields.
Closes #772 

PlantHydraulicsParameters previously held a root distribution function
that had a default value for rooting depth. This function is
always the same outside testing, so it was removed as a parameter
and placed into src. rooting_depth, an argument to the distribution function
is now a parameters that is either a Float or a field. The tests are changed to
check passing rooting_depth as a float or a field of floats that are the same everywhere.

## To-do
- Add to ClimaArtifacts

## Content
- Rooting_depth map created
- rooting_depth added as parameter
- rooting_distribution moved to source
- all calls to PlantHydraulicsParameters changed accordingly 
- tests changed to also test rooting_depth as a field
- add message in NEWS.md

PlantHydraulicsParameters previously held a root distribution function
that had a default value for rooting depth. This function is
always the same outside testing, so it was removed as a parameter
and placed into src. rooting_depth, an argument to the distribution function
is now a parameters that is either a Float or a field. The tests are changed to
check passing rooting_depth as a float or a field of floats that are the same everywhere.

The rooting_depth map can be seen here:

rooting_depth

----
- [x] I have read and checked the items on the review checklist.

### CODE REVIEW

File: NEWS.md
Comment (Sbozzolo) [inline NEWS.md:11]: If something is breaking, it is always a good idea to describe how it is breaking and what users might need to do to fix it

File: test/standalone/Vegetation/plant_hydraulics_test.jl
Comment (Sbozzolo) [inline test/standalone/Vegetation/plant_hydraulics_test.jl:332]: Let's discuss this. I would have expected `F` to be a `Field` here too

Comment (kmdeck) [reply test/standalone/Vegetation/plant_hydraulics_test.jl:332]: I believe that in this test, we solve for the steady state of a vector-valued quantity using a root solver. We then set the initial condition of the field to the value of this vector. This seemed easier than getting the root solver to play with Fields, though maybe that is easy enough to do

File: src/standalone/Vegetation/PlantHydraulics.jl
Comment (kmdeck) [inline src/standalone/Vegetation/PlantHydraulics.jl:377]: Could you sketch out here the connection between rooting depth and beta?
E.g. give the CDF F(z) = 1-beta^(100z) and cite Equation 2.23 of Bonan, "Climate Change and Terrestrial Ecosystem Modeling", 2019 Cambridge University Press. And say this is equivalent to this probability density function with coordinate `z` replacing depth.

File: src/standalone/Vegetation/PlantHydraulics.jl
Comment (kmdeck) [inline src/standalone/Vegetation/PlantHydraulics.jl:380]: If `z` and `rooting_depth` are already FT, and we change 1.0 -> 1, do we still need the FTs in the formula?

File: src/standalone/Vegetation/PlantHydraulics.jl
Comment (kmdeck) [inline src/standalone/Vegetation/PlantHydraulics.jl:152]: Maybe:
"Rooting depth parameter (m) - a characteristic depth below which `1/e` of the root mass lies."

File: src/standalone/Vegetation/PlantHydraulics.jl
Comment (imreddyTeja) [inline src/standalone/Vegetation/PlantHydraulics.jl:386]: @kmdeck Is this clear and accurate? I was not sure if there is a citation convention to follow. 

In your comment you said to "say this is equivalent to this probability density function with coordinate z replacing depth," but the z coordinate in the CDF is still present in the PDF, just negated. 

Could you sketch out here the connection between rooting depth and beta?
E.g. give the CDF F(z) = 1-beta^(100z) and cite Equation 2.23 of Bonan, "Climate Change and Terrestrial Ecosystem Modeling", 2019 Cambridge University Press. And say this is equivalent to this probability density function with coordinate z replacing depth.

Comment (kmdeck) [reply src/standalone/Vegetation/PlantHydraulics.jl:386]: I was making a distinction between "depth" z as Bonan uses it, and the coordinate z, what you have is clear! I would just check for typos ("probabilty" twice), replace beta with \beta (the symbol).

thank you!

File: test/standalone/Vegetation/plant_hydraulics_test.jl
Comment (imreddyTeja) [reply test/standalone/Vegetation/plant_hydraulics_test.jl]: @Sbozzolo I changed this test to not take the mean, and instead follow the original behavior but with fields. My implementation of this change is abnormal (see last commit). Could we discuss?

File: src/standalone/Vegetation/PlantHydraulics.jl
Comment (Sbozzolo) [inline src/standalone/Vegetation/PlantHydraulics.jl:195]: You are doing manual type dispatch here. Instead, you can just define two methods for the function to return the correct value.

File: test/standalone/Vegetation/plant_hydraulics_test.jl
Comment (Sbozzolo) [inline test/standalone/Vegetation/plant_hydraulics_test.jl:16]: This is no longer needed

File: test/standalone/Vegetation/plant_hydraulics_test.jl
Comment (Sbozzolo) [inline test/standalone/Vegetation/plant_hydraulics_test.jl:338]: Can you add a description of what this test represents/is doing?

File: test/standalone/Vegetation/plant_hydraulics_test.jl
Comment (Sbozzolo) [reply test/standalone/Vegetation/plant_hydraulics_test.jl]: It's a little abnormal, but I think it's fine! Thanks for the comments in the commit message too

File: src/standalone/Vegetation/PlantHydraulics.jl
Comment (imreddyTeja) [reply src/standalone/Vegetation/PlantHydraulics.jl:195]: Julia doesn't use keyword arguments as dispatch. When I tried defining two methods for the function, the second just overwrote the first. It seems like this was the solution reached here:
https://discourse.julialang.org/t/how-can-kwargs-be-used-in-multiple-dispatch/86790

I could make the keyword arguments positional, but that would make the change breaking. Do you see a better way to do this, or should I revert to making this a breaking change?

Comment (Sbozzolo) [reply src/standalone/Vegetation/PlantHydraulics.jl:195]: It's fine! I didn't pay attention to `root_distribution` being a kwarg. Thank you!

File: src/standalone/Vegetation/PlantHydraulics.jl
Comment (charleskawczynski) [inline src/standalone/Vegetation/PlantHydraulics.jl:189]:
```suggestion
        rooting_depth = if root_distribution isa Nothing
            rooting_depth
        else
            FT(1.0 / root_distribution(0.0))
        end
        return PlantHydraulicsParameters{
            FT,
            typeof(ai_parameterization),
            typeof(conductivity_model),
            typeof(retention_model),
            typeof(rooting_depth),
        }(
            ai_parameterization,
            ν,
            S_s,
            conductivity_model,
            retention_model,
            rooting_depth,
        )
```
Couldn't we write this like this instead?

File: test/standalone/Vegetation/plant_hydraulics_test.jl
Comment (charleskawczynski) [inline test/standalone/Vegetation/plant_hydraulics_test.jl:289]: `FT` here should probably not be grabbed from the outer-scope, as that will result in a captured variable, and that can lead to type instabilities (which can hurt performance).

Comment (imreddyTeja) [reply test/standalone/Vegetation/plant_hydraulics_test.jl:289]: I'm not sure how this test could be rewritten to not use variables from the outer scope. Would passing in the outer scope variables as arguments fix the issue?

`function create_exp_tendency_func(rooting_depth, LAI, FT)
                # Solve for hydrostatic equilibrium
                    function initial_compute_exp_tendency!(F, Y)
                        AI = (; leaf = LAI(1.0), root = RAI, stem = SAI)
                        T0A = FT(1e-8) * AI[:leaf]`

When I check using @code_warntype, both create_exp_tendency_func and the function it returns are type stable.  Doesn't the inner function still capture the variables FT and LAI though?

Comment (kmdeck) [reply test/standalone/Vegetation/plant_hydraulics_test.jl:289]: I think in this case it is fine to use variables from the outer scope, as this function is run 1x, and in a test where performance is not important.

File: src/standalone/Vegetation/PlantHydraulics.jl
Comment (kmdeck) [inline src/standalone/Vegetation/PlantHydraulics.jl:193]: Not to stir the pot, I know there was just a discussion about breaking changes, but I dont think we should support passing in the root_distribution function. Or, if we do, we need to know for that distribution how to compute the rooting_depth according to our fixed, single, rooting distribution in src (as of this PR). We cant assume that the rooting distribution they provide has the same functional form, from which we know how to compute the depth

**Comment by kmdeck** [comment]
@imreddyTeja this PR looks good to me - I think you are you able to resolve some of the comments above? Are you waiting on us for approval?

**Comment by imreddyTeja** [comment]
I was making it a non-breaking change again. It is ready for approval review now.

**Review by kmdeck — APPROVED** [review]
I appreciate the care you took with this, thank you!

**Review by Sbozzolo — APPROVED** [review]

**Pull request merged**
